006-restarting-kafka-connect-connectors-and-tasks.md

```diff
@@ -0,0 +1,137 @@
+
+When applied to a `KafkaConnector` CR, the annotation value contains the ID of the task that is restarted. For example:
+```
+apiVersion: strimzi.io/v1alpha1
+kind: KafkaConnector
+metadata:
+  name: my-source-connector
+  annotations:
+    strimzi.io/restart-task: "0"
+  ...
+```
+
+### `strimzi.io/restart-connector-task` annotation
+
+An annotation named `strimzi.io/restart-connector-task` can be applied to the `KafkaMirrorMaker2` custom resource. The value of the annotation and the CR it is applied to determines the connector task that is restarted. Once the operator successfully calls the restart REST API endpoint, the CR is patched to remove the annotation. If the restart REST API endpoint does not return a successful (HTTP 200 OK) response, the operator will continue to attempt to call the restart REST API endpoint once per reconcile cycle.
+
+When applied to a `KafkaMirrorMaker2` CR, the annotation value contains the name of the connector and the ID of the task that is restarted, delimited by a `:` character. For example:
+
+```
+apiVersion: strimzi.io/v1alpha1
+kind: KafkaMirrorMaker2
+metadata:
+  name: my-mm2-cluster
+  annotations:
+    strimzi.io/restart-task: "source-cluster->target-cluster.MirrorSourceConnector:2"
+  ...
+```
+
+
+### Automatically restarting `FAILED` connectors and tasks
+
+The `KafkaConnector` and `KafkaMirrorMaker2` operators will have new default behaviour that automatically restarts connectors or tasks that are in the `FAILED` state. For each connector or task that is in the `FAILED` state, the operator will call the Connect REST API `restart` endpoint once per reconcile cycle to attempt to restart the connector/task.
+
+To disable the new default behaviour a new field in the `KafkaConnector` spec named `autoRestart` with a value of `false` can be applied. This will be useful when users know that external factors will cause connectors/tasks to be in the `FAILED` state and they do not want the operator attempting automatic restarts. It also means users can keep the current behaviour of the `KafkaConnector` and `KafkaMirrorMaker2` operators if they wish.
+
+For example, the following `KafkaConnector` CR disables the automatic restarting of the connector and it's tasks when they are in the `FAILED` state:
+```
+apiVersion: strimzi.io/v1alpha1
+kind: KafkaConnector
+metadata:
+  name: my-source-connector
+spec:
+  autoRestart: "false"
+  ...
+```
+
+Similarly, a new field in the `KafkaMirrorMaker2` spec named `autoRestartConnectorsAndTasks` with a value of `false` can be applied, to disable auto-restarting connectors and tasks.
```

Comment:
I'm not sure I understand why does `KafkaConnector` have `autoRestart` while `KafkaMirrorMaker2` has `autoRestartConnectorsAndTasks`. It seems to do the same in both cases - restart both connectors and tasks - so why not use the same name?

Reply:
I did this to go along with @ppatierno's comment on the annotations (https://github.com/strimzi/proposals/pull/8#discussion_r453168363). It was felt that the `strimzi.io/restart` and `strimzi.io/restart-task` annotations made sense on a KafkaConnector resource because it's clear that users are asking to restart the resource, but it wasn't so clear on a KafkaMirrorMaker2 resource where the intention is to restart the MM2 connectors - hence I updated the annotations to `strimzi.io/restart-connector` and `strimzi.io/restart-task` for KafkaMirrorMaker2 CRs.
When I added the `autoRestart` field to the KafkaConnector spec, I expected the same issue would apply, so I named it `autoRestartConnectorsAndTasks` for KafkaMirrorMaker2. I'm happy to give them the same name if people agree - @ppatierno what do you think?

Reply:
I guess I can see that angle - yeah, it might be confusing if you don't know what it does. So I think I'm ok with the current naming. Thanks for explaining.